### ISSUE TITLE
dependencies with "tag || version" are skipped.

### ISSUE THREAD

**Issue by jsg2021** [npm/npm#16931]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [ ] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?

Dependencies with version strings of `"tagname || ^1.0.0"` are completely skipped. I believe this worked in 4. The goal for this is for a CI snapshot system ... as app libraries are developed, they are snapshot and published under a tag, say "next" or "alpha"... the main app's package.json is modified by the CI to change all internal deps version strings to be prefixed with `snapshot-tag-name || ` and  the app's snapshot is then generated with snapshot dependencies, or fallback to semvar.

#### How can the CLI team reproduce the problem?

run install with a dependency that has a version string like "next || ^1.0.0". (the tag name doesn't have to exist, same result either way)

### supporting information:

 - `npm -v` prints: 5.0.1
 - `node -v` prints: 8.0.0
 - `npm config get registry` prints: private-sinopia-url
 - Windows, OS X/macOS, or Linux?: macOS
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by zkat** [comment]
Closing as dupe of https://github.com/npm/npm/issues/16744

**Comment by jsg2021** [comment]
is the title of the other ticket a misnomer? Was the sub-dependency's version specifier the real issue? I think my issue has a valid version string... the documents do say I can `or` values?

**Comment by zkat** [comment]
@jsg2021 I marked it as a duplicate because there's something fishy where certain types of dep specifiers that apparently used to work aren't supported. `tag || semver` isn't a valid semver range, and I think it's a good clue about what's going on (for us).

**Comment by jsg2021** [comment]
really `tag || semver` isn't valid? bummer :(

**Comment by zkat** [comment]
not as a _semver_ identifier. But then again, neither are tags. So something fishy's going on.

**Comment by zkat** [comment]
You know what, since this one deals specifically with tags, I'm going to reopen this, cause idk what changed to cause this.

**Comment by jsg2021** [comment]
ha ok :)

**Comment by zkat** [comment]
@jsg2021 https://github.com/npm/npm/pull/17007

I've no idea how we stopped parsing these in loose mode, but there ya have it. That patch'll fix it once it lands. It'll probably be released with `npm@5.0.4` sometime this week.

**Comment by jsg2021** [comment]
thanks :)

**Comment by jsg2021** [comment]
I would have been satisfied if you had told me it was invalid :) 

When I read the docs, I inferred the `||` operator allowed for any version specifier in each operand... but if it was intended to be strictly version/ranges that would be okay too... I'd just change my CI plan :)

**Comment by zkat** [comment]
@jsg2021 It's still invalid! But it will *work*. There's a separate issue now where someone offered to make a PR to actually warn about these when normalizing package json.

Since this is back to working how it used to in the latest canary, I'm gonna go ahead and close this. Thanks again for the report! 🎉
